python/ray/tune/result.py

```diff
@@ -2,101 +2,120 @@
-        "date",
-
-        # (Auto-filled) A UNIX timestamp of when the result was processed.
-        "timestamp",
-
-        # (Auto-filled) The hostname of the machine hosting the
-        # training process.
-        "hostname",
-
-        # (Auto-filled) The node ip of the machine hosting the
-        # training process.
-        "node_ip",
-
-        # (Auto=filled) The current hyperparameter configuration.
-        "config",
-    ])
 
-TrainingResult.__new__.__defaults__ = (None, ) * len(TrainingResult._fields)
+class TrainingResult(MutableMapping):
+    """Dict for storing results for Tune, with extra properties.
+
+    When using Tune with custom training scripts, you must periodically report
+    training status back to Ray by calling reporter(result).
+
+    Properties:
+        done (bool): (Optional) If training is terminated.
+        time_this_iter_s (float): (Auto-filled) Time in seconds
+            this iteration took to run. This may be overriden in order to
+            override the system-computed time difference.
+        time_total_s (float): (Auto-filled) Accumulated time in seconds
+            for this entire experiment.
+        experiment_id (str): (Auto-filled) Unique string identifier
+            for this experiment. This id is preserved
+            across checkpoint / restore calls.
+        training_iteration (int): (Auto-filled) The index of this
+            training iteration, e.g. call to train().
+        pid (str): (Auto-filled) The pid of the training process.
+        date (str): (Auto-filled) A formatted date of
+            when the result was processed.
+        timestamp (str): (Auto-filled) A UNIX timestamp of
+            when the result was processed.
+        hostname (str): (Auto-filled) The hostname of the machine
+            hosting the training process.
+        node_ip (str): (Auto-filled) The node ip of the machine
+            hosting the training process.
+    """
+    def __init__(self, *args, **kwargs):
+        # `done` has special treatment because the responsibility
+        # to set it is shared between user and system, and to ensure
+        # CSV Writer doesn't fail, this is defaulted just in case.
+        self._store = dict(done=False)
+        self.update(dict(*args, **kwargs))
+
+    def __getitem__(self, key):
+        return self._store[key]
+
+    def __setitem__(self, key, value):
+        self._store[key] = value
+
+    def __delitem__(self, key):
+        del self._store[key]
+
+    def __iter__(self):
+        return iter(self._store)
+
+    def __len__(self):
+        return len(self._store)
+
+    def copy(self):
+        return TrainingResult(**self._store)
+
+    def __repr__(self):
+        return "TrainingResult({})".format(self._store)
+
+    def as_dict(self):
+        """Retrieves copy ofinternal dict, used for JSON dumping."""
+        return copy.deepcopy(self._store)
+
+    @property
+    def done(self):
+        return self.get("done")
```

Comment:
You can replace all of these with a `__getattr__`

Reply:
So the intent here was not to generalize the namedtuple functionality.

I think there are a small number of fields that should be special-cased, specifically the fields set and used by other parts of Tune. Everything else that the user touches should be only accessible via getitem/setitem syntax. The reasoning for this is that as a user, intuitively you would expect reading custom attributes and writing custom attributes to have the same semantics.

Is the `__getattr__` after subclassing dict standard?

Reply:
So presumably if you replace `__getattr__`, you'll also want to set `__setattr__` to raise an exception if called?